velox/functions/prestosql/fuzzer/TruncateArgGenerator.h

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/expression/fuzzer/DecimalArgGeneratorBase.h"
+
+namespace facebook::velox::exec::test {
+
+class TruncateArgGenerator : public fuzzer::ArgGenerator {
+ public:
+  std::vector<TypePtr> generateArgs(
+      const exec::FunctionSignature& signature,
+      const TypePtr& returnType,
+      FuzzerGenerator& rng) override {
+    // Only the single-arg truncate function is supported because
+    // ArgumentTypeFuzzer can generate argument types for the two-arg truncate
+    // function.
+    VELOX_CHECK_EQ(1, signature.argumentTypes().size());
+    // Generates a decimal type following below formulas:
+    // p = max(p1 - s1, 1)
+    // s = 0
+    const auto [p, s] = getDecimalPrecisionScale(*returnType);
+    if (s != 0) {
+      return {};
+    }
+
+    // p1 = s1 or p1 = s1 + 1.
```

Comment:
I'm having a hard time understanding this code. Would you help explain this logic?

I'm thinking that there are 2 cases: (1) p == 1, (2) p > 1.

When p == 1, input can be (s1, s1) with s1 in [1, 38] or (1 + s1, s1) with s1 in [0, 37]. The latter case is the same as when p > 1.

When p > 1, input can be (p + s1, s1) with s1 in [0, 38 - p].

Hence, we could have 2 branches:

1. p == 1 and coinToss() == true => s1 = rand(1, 38); return (s1, s1);
2. s1 = rand(0, 38 - p); return (p + s1, s1);

Reply:
```
const auto s1 = rand32(38, rng);
const auto p1 = s1 == 0 ? 1 : std::min(s1 + rand32(1, rng), (uint32_t)38);
```
I apologize for confusion brought by above logic. I was thinking there were two cases when p == 1: p1 = s1 or p1 = s1 + 1, and was attempting to handle both in one branch. Basically, `p1 = s1 + rand32(1, rng)` since `1` is optional and `rand32(1, rng)` returns 0 or 1. But p1 cannot exceed 38, so std::min(xx, 38) is used to ensure that. In particular, p1 can only be 1 when s1 is 0.

Thank you so much for sharing your thoughts. It is much clearer to handle the  `p1 = s1 + 1` case together with `p > 1`. I revised this logic as your suggestion.

Reply:
@rui-mo Rui, thank you for explaining. Makes sense to me now.